src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -641,16 +596,118 @@ private class Emit(
     }
   }
 
+  private def emitArrayIterator(ir: IR, env: E): (Code[Unit], Option[Code[Int]], F => (Code[Unit], Code[Boolean], Code[Unit])) = {
+
+    def emit(ir: IR, env: E = env) = this.emit(ir, env)
+
+    def emitArrayIterator(ir: IR, env: E = env) = this.emitArrayIterator(ir, env)
+
+    val region = fb.getArg[Region](1).load()
+
+    ir match {
+      case x@ArrayRange(startir, stopir, stepir) =>
+        val (d, m, v) = Array(emit(startir), emit(stopir), emit(stepir)).unzip3
+        val start = fb.newLocal[Int]("ar_start")
+        val stop = fb.newLocal[Int]("ar_stop")
+        val step = fb.newLocal[Int]("ar_step")
+
+        val i = fb.newLocal[Int]("ar_i")
+        val len = fb.newLocal[Int]("ar_len")
+        val llen = fb.newLocal[Long]("ar_llen")
+
+        val calcLength = Code(
+          start := coerce[Int](v(0)),
+          stop := coerce[Int](v(1)),
+          step := coerce[Int](v(2)),
+          step.ceq(0).mux(
+            Code._fatal("Array range cannot have step size 0."),
+            Code._empty[Unit]),
+          llen := start.ceq(stop).mux(0L,
+            (step < 0).mux(
+              (start.toL - stop.toL - 1L) / (-step).toL + 1L,
+              (stop.toL - start.toL - 1L) / step.toL + 1L)),
+          (llen > const(Int.MaxValue.toLong)).mux(
+            Code._fatal("Array range cannot have more than MAXINT elements."),
+            len := (llen < 0L).mux(0L, llen).toI)
+        )
+
+        (calcLength ,Some(len.load()), { continuation: F =>
+          (coerce[Unit](Code(d: _*)), m.reduce(_ || _), Code(
+            i := 0,
+            Code.whileLoop(i < len,
+              continuation(false, start),
+              i := i + 1,
+              start := start + step)))
+        })
+      case x@ArrayFilter(a, name, condition) =>
+        val elementTypeInfoA = coerce[Any](typeToTypeInfo(x.typ.elementType))
+        val xmv = mb.newBit()
+        val xvv = fb.newLocal(name)(elementTypeInfoA)
+        val condenv = env.bind(name -> (elementTypeInfoA, xmv, xvv))
+        val (docond, mcond, vcond) = emit(condition, condenv)
+
+        val filterCont = { (cont: F, m: Code[Boolean], v: Code[_]) =>
+          Code(
+            xmv := m,
+            xvv := v,
```

Comment:
This doesn't seem safe, we shouldn't look at `v` unless `m` is `false`. Am I misremember how this works? I think this would be the right thing:

```scala
Code(
  xmv := m,
  xmv.mux(
    Code._empty,
    Code(
      xvv := v,
      docond,
      (mcond || !coerce[Boolean](vcond)).mux(
        Code._empty,
        cont(false, xvv)))))
```

Reply:
cc @catoverdrive

Reply:
ah, you're totally right.